src/pybind/mgr/smb/resources.py

```diff
@@ -425,19 +447,27 @@ def _customize_resource(rc: resourcelib.Resource) -> resourcelib.Resource:
 ]
 
 
-def load_text(blob: str) -> List[SMBResource]:
+def load_text(
+    blob: str, *, input_sample_max: int = 1024
+) -> List[SMBResource]:
     """Given JSON or YAML return a list of SMBResource objects deserialized
     from the input.
     """
+    json_err = None
     try:
-        data = yaml.safe_load(blob)
-    except ValueError:
-        pass
-    try:
+        # apparently JSON is not always as strict subset of YAML
+        # therefore trying to parse as JSON first is not a waste:
+        # https://john-millikin.com/json-is-not-a-yaml-subset
         data = json.loads(blob)
-    except ValueError:
-        pass
-    return load(data)
+    except ValueError as err:
+        json_err = err
+    try:
+        data = yaml.safe_load(blob) if json_err else data
+    except (ValueError, yaml.parser.ParserError) as err:
+        raise InvalidInputError(str(err), blob) from err
+    if not isinstance(data, (list, dict)):
+        raise InvalidInputError("input must be an object or list", blob)
+    return load(cast(Simplified, data))
```

Comment:
We try json first because json is stricter and more likely to fail. Originally, I had hoped to only need YAML parsing but I found a couple of resources that stated that yaml and json do differ in a few places and so while YAML is almost a superset of JSON it isn't quite so.

So we try JSON first, expecting it to fail quickly if the input is not JSON and then retry with YAML as it has the more flexible syntax.

Correct wrt to the spec. Once we have sourced the YAML/JSON into a python dict it will be additionally validated by the resources/resourcelib.py code.

Reply:
There's a link in the nearby comment WRT JSON not being a subset of YAML.